### PR TITLE
Allow user-defined codecs

### PR DESCRIPTION
**Pull request by aleixalcacer**
Experimental work

### CODE REVIEW

File: blosc/blosc2.h [hunk 1]
```diff
@@ -774,7 +796,7 @@ typedef struct {
  * @brief Default struct for compression params meant for user initialization.
  */
 static const blosc2_cparams BLOSC2_CPARAMS_DEFAULTS = {
-        BLOSC_BLOSCLZ, 5, 0, 8, 1, 0, NULL,
+        BLOSC_BLOSCLZ, 0,{0}, 5, 0, 8, 1, 0, NULL,
```
Comment (FrancescAlted) [inline blosc/blosc2.h:799]: Missing space.

File: blosc/blosc2.h [hunk 2]
```diff
@@ -789,12 +811,14 @@ typedef struct {
...
   //!< The associated schunk, if any (NULL).
+  blosc2_udcodec udcodecs[BLOSC2_MAX_UDCODECS];
+  //!< The user-defined compression codecs.
 } blosc2_dparams;
 
 /**
  * @brief Default struct for decompression params meant for user initialization.
  */
-static const blosc2_dparams BLOSC2_DPARAMS_DEFAULTS = {1, NULL};
+static const blosc2_dparams BLOSC2_DPARAMS_DEFAULTS = {1, NULL,{0}};
```
Comment (FrancescAlted) [inline blosc/blosc2.h:821]: Missing space.

File: examples/udcodecs.c [hunk 3]
```diff
@@ -0,0 +1,180 @@
+/*
+  Copyright (C) 2015  Francesc Alted
```
Comment (FrancescAlted) [inline examples/udcodecs.c:2]: Update copyright

File: README_CHUNK_FORMAT.rst [hunk 4]
```diff
@@ -125,6 +125,11 @@ for encoding blocks with a filter pipeline::
     to their index in increasing order. The type of filter applied is specified by the `filter_code`. Each
     `filter_code` has an associated field in `filter_meta` that can contain metadata about the filter.
 
+:compcode_meta:
+    (``uint8``) Compression codec metadata.
+
+    Possible metadata associated with the compression codec.
```
Comment (FrancescAlted) [inline README_CHUNK_FORMAT.rst:131]: Better rephrase this as:  "Metadata associated to User Defined Codec.  0 if not User Defined Codec."

File: blosc/blosc2.h (hunk 1 above)
Comment (aleixalcacer) [reply blosc/blosc2.h]: Fixed

File: blosc/blosc2.h (hunk 2 above)
Comment (aleixalcacer) [reply blosc/blosc2.h]: Fixed

File: examples/udcodecs.c (hunk 3 above)
Comment (aleixalcacer) [reply examples/udcodecs.c]: Fixed

File: README_CHUNK_FORMAT.rst (hunk 4 above)
Comment (aleixalcacer) [reply README_CHUNK_FORMAT.rst]: I rephrase the sentence as: "Possible metadata associated with the compression codec. Only used in user-defined codecs."

Additionally, a sentence like "See more here" can be added when a README_PLUGINS.md is redacted.

What do you think?

Comment (FrancescAlted) [reply README_CHUNK_FORMAT.rst]: Ok. I'd just remove the 'Possible' which does not bring siginificant meaning.

Comment (aleixalcacer) [reply README_CHUNK_FORMAT.rst]: Fixed

File: blosc/blosc2.c [hunk 5]
```diff
@@ -3100,7 +3176,9 @@ const char* blosc_get_compressor(void)
 
 int blosc_set_compressor(const char* compname) {
   int code = blosc_compname_to_compcode(compname);
-
+  if (code >= BLOSC_UDCODEC) {
```
Comment (FrancescAlted) [inline blosc/blosc2.c:3179]: Please add a `BLOSC_TRACE_ERROR()` with the reason.  E.g. "User defined codecs cannot be set here.  Use Blosc2 mechanism instead."

File: blosc/blosc2.c [hunk 6]
```diff
@@ -3724,3 +3833,32 @@ int blosc2_chunk_repeatval(blosc2_cparams cparams, const size_t nbytes,
...
+      return BLOSC2_ERROR_CODEC_PARAM;
+    }
+  }
+
+  blosc2_codec *codec_new = malloc(sizeof(blosc2_codec));
+  memcpy(codec_new, codec, sizeof(blosc2_codec));
+  g_codecs[g_ncodecs++] = codec_new;
+
+  return BLOSC2_ERROR_SUCCESS;
+}
```
Comment (FrancescAlted) [inline blosc/blosc2.c:3864]: Add a blank line at the end.

File: blosc/blosc2.h [hunk 7]
```diff
@@ -154,9 +155,19 @@ enum {
   BLOSC_SNAPPY = 3,
   BLOSC_ZLIB = 4,
   BLOSC_ZSTD = 5,
-  BLOSC_MAX_CODECS = 7,  //!< maximum number of reserved codecs
+  BLOSC_UDCODEC = 6,
+  BLOSC_LAST_CODEC = 7,  //!< sentinel
+  BLOSC_LAST_REGISTERED_CODEC = 32,  //!< sentinel
```
Comment (FrancescAlted) [inline blosc/blosc2.h:160]: I'd redefine this as:

```
BLOSC_LAST_REGISTERED_CODEC = BLOSC2_BDEFINED_CODECS + 0
```

and add a comment explaining that this number (0) should be increased as we register new codecs.

**Review by FrancescAlted — APPROVED** [review]
Looks good to me!  I just have some simple comments.
